src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java

```diff
@@ -0,0 +1,124 @@
+     * @return the persisted entity
+     */
+    public Organization save(Organization organization) {
+        log.debug("Request to save Organization : {}", organization.getName());
+        return organizationRepository.save(organization);
+    }
+
+    /**
+     * Add a new organization and execute indexing based on its emailPattern
+     * @param organization the organization to add
+     * @return the persisted organization entity
+     */
+    public Organization add(Organization organization) {
+        Organization addedOrganization = save(organization);
+        addedOrganization = organizationRepository.findOneWithEagerUsersAndCoursesOrElseThrow(addedOrganization.getId());
+        indexing(addedOrganization);
+        return addedOrganization;
+    }
+
+    /**
+     * Update an organization
+     * To avoid removing the currently mapped users and courses of the organization,
+     * these are loaded eagerly and the edited values changed within the loaded entity.
+     * @param organization the organization to update
+     * @return the updated organization
+     */
+    public Organization update(Organization organization) {
+        log.debug("Request to update Organization : {}", organization);
+        boolean indexingRequired = false;
+        Optional<Organization> optionalOldOrganization = organizationRepository.findByIdWithEagerUsersAndCourses(organization.getId());
+        if (optionalOldOrganization.isPresent()) {
+            Organization oldOrganization = optionalOldOrganization.get();
+            if (!oldOrganization.getEmailPattern().equals(organization.getEmailPattern())) {
+                indexingRequired = true;
+            }
+            oldOrganization.setName(organization.getName());
+            oldOrganization.setShortName(organization.getShortName());
+            oldOrganization.setUrl(organization.getUrl());
+            oldOrganization.setDescription(organization.getDescription());
+            oldOrganization.setLogoUrl(organization.getLogoUrl());
+            oldOrganization.setEmailPattern(organization.getEmailPattern());
+            if (indexingRequired) {
+                indexing(oldOrganization);
+            }
+            return organizationRepository.save(oldOrganization);
+        }
+        else {
+            throw new EntityNotFoundException("Organization with id: \"" + organization.getId() + "\" does not exist");
+        }
+    }
+
+    /**
+     * Delete an organization
+     * @param organizationId the id of the organization to delete
+     */
+    public void deleteOrganization(Long organizationId) {
+        final Organization organization = organizationRepository.findOneWithEagerUsersAndCoursesOrElseThrow(organizationId);
+
+        // we make sure to delete all relations before deleting the organization
+        organization.getUsers().forEach(user -> {
+            userRepository.removeOrganizationFromUser(user.getId(), organization);
```

Comment:
Not sure here but since it's a database call, would that impact performance if the organization has many users? Is there a way to batch-delete entities?

Reply:
Tried to come up with a native query removing the relative relationships but does not seem to work.. 
Unfortunately I did not find an alternative solution and kept these for loops as done by removing courses (`CourseService.java - public void delete(Course course)`)